### DISCUSSION TITLE
Migrate to an alternative branching strategy

### DISCUSSION THREAD

**Discussion by scroix** [museumsvictoria/nodel#316] [General]
@justparking There are a bunch of open-source projects with limited maintainers that serve as an example for good branching strategy.

https://github.com/mrdoob/three.js
- 100k stars, 35k forks, 40k commits
- 1 - 2 maintainers

https://github.com/tailwindlabs/tailwindcss/
- 207k stars, 34k forks, 3.5k commits
- 1 maintainer

https://github.com/vuejs
- 78k stars, 3.9k forks, 5k commits
- 1-2 maintainers

The strategy is essentially the same in all three of them.
- A `main` or `master` branch contains the latest stable release
- A `dev` or `next` branch for features and bug features by external contributors; _crucially_, this is often the **default** branch.
- There are sometimes `release` branches to be used as a staging for final testing (e.g. 24-hours soap dish timing based tests), bug fixes and a chance to update documentation before going to `main`. The benefit, I assume, would be to allow continued development whilst locking off the latest commits from `dev` before a release.

**Comment by justparking** [comment]
I think that'll work well.

Have you consider how we might work on a particular feature/update **collectively**?

For example, your PR - https://github.com/scroix/nodel/tree/feature/introduce-junit-testing - if I found a little typo, how would I go about correcting it?

I just tried to update your branch and hit this obvious hurdle:

![image](https://github.com/museumsvictoria/nodel/assets/5266533/e674eaa1-96d1-47de-bbed-cf1c840217e3)

So that can work i.e. a fork is done and then another PR can be created, however means all updates have to be done via the PRs no matter how big or small the update.

Is there any other way of doing this without moving from user-based repos to central ones: e.g.
* `:scroix/nodel/introduce-junit-testing` vs 
* `:museumsvictoria/nodel-dev/introduce-junit-testing`?

**Reply by scroix** [reply]
I'm not sure a dev-repo is necessary. You _should_ be able to edit these PRs.

![image](https://github.com/museumsvictoria/nodel/assets/9277107/9c9a92c1-1902-48dc-ae7f-e787c3e2f4b3)

It's a little weird though - the PR exists as separate branch. You can either

1. Check it out with a specific reference to the PR origin `git fetch origin pull/$ID/head:$BRANCHNAME
` ([as described here](https://stackoverflow.com/questions/27567846/how-can-i-check-out-a-github-pull-request-with-git))
2. Use GitHub's CLI/GUI to check it out. `gh pr checkout $ID` ([as described here](https://docs.github.com/en/enterprise-server@3.9/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally))

![image](https://github.com/museumsvictoria/nodel/assets/9277107/144425d5-2016-4ab5-870b-0dc5c5377cec)

**Reply by justparking** [reply]
OK, no worries. I think I was able to update a PR of yours via the git tools and it actually worked. Perhaps my method of doing it directly in github treats fork/branch permissions differently.

**Comment by justparking** [comment]
I've just implemented this:

> A main or master branch contains the latest stable release
> A dev or next branch for features and bug features by external contributors; crucially, this is often the default branch.

No when the repo is cloned, `dev` is the default.

And new PRs will target `dev`

**Comment by scroix** [comment]
I see this - It looks _good_. There was a bunch of old branches pruned too! 💛 

I went ahead and updated my pending PRs to reference `master` as the "release" branch when building binaries as well (https://github.com/museumsvictoria/nodel/pull/313/commits/15b641654cb971901b35d9929f54c88c57351bcf).
